### DISCUSSION TITLE
Use same transaction in different methods

### DISCUSSION THREAD

**Discussion by aleripe** [powersync-ja/sqlite_async.dart#16] [Q&A] [answered]
Hi guys,
forst of all thanks for this great library.

I want to expose you my problem that might already have a solution, I don't know. Let's suppose I have two methods:

```
Future<void> methodA() async {
    await _dbContext.executeInTransaction(() async {
       // (insert, update, delete)
   }
}

Future<void> methodB() async {
    await _dbContext.executeInTransaction(() async {
       await methodA();
       // (insert, update, delete)
   }
}
```
`executeInTransaction` in `methodA` should be able to issue a new transaction if it is not available yet , or reuse the transaction created by `methodB` in the second case.

I have only found the method `writeTransaction` which correctly offers a transaction object but then commits it at the end of the callback, so it's not appropiate for my case as `methodB` with `methodA` inside would find it closed.

I don't know if this makes sense to you. Is there way to obtain this behavior with your library?

Thanks!

**Comment by rkistner** [comment]
That's a good question.

What I typically do is pass the transaction object along to any other methods that need it. In your example, it would look like this:

```dart
Future<void> methodA(SqliteWriteContext tx) async {
   // (insert, update, delete) on tx
}

Future<void> methodB() async {
    await _dbContext.executeInTransaction((tx) async {
       await methodA(tx);
       // (insert, update, delete)
   }
}
```

If you also need `methodA` to start a transaction, I'd split it into two:

```dart
Future<void> methodA() async {
    await _dbContext.executeInTransaction((tx) async {
       await methodAWithTx(tx);
   }
}

Future<void> methodAWithTx(SqliteWriteContext tx) async {
   // (insert, update, delete) on tx
}

Future<void> methodB() async {
    await _dbContext.executeInTransaction((tx) async {
       await methodAWithTx(tx);
       // (insert, update, delete)
   }
}
```

There isn't a nice way currently to automatically start a transaction when not in one yet (there may be a workaround), but I've found in most cases it's better to be explicit about whether a method is called inside a transaction, or starts its own transaction.

**Reply by aleripe** [reply]
Thanks for your kind answer! I considered that but I have thousands of login business methods and I didn't want to clutter them passing the `Transaction` object everywhere. Is there a way to obtain a `Transaction` and use it manually without resorting to the scope? Thanks again!

**Reply by rkistner** [reply]
I've never tried other approaches, but two that may work:

1. Set the transaction as a value on the zone, using e.g. using [runZoned](https://api.flutter.dev/flutter/dart-async/runZoned.html), which would then be accessible to any code running within that zone.
2. Store the current transaction as a static/global variable. A little simpler than using zones, but you'd need to be careful to avoid concurrency issues.

**Reply by aleripe** [reply]
I've yet to find a way to obtain that `Transaction`, on `SqliteDatabase` I see only `readTransaction` and `writeTransaction` that are scoped and would automatically close the transaction at the end of the scope. Isn't there a `getTransaction` or `openTransaction` somewhere? :)

Thanks again for your kind help!

**Reply by rkistner** [reply]
If you just want to avoid passing the transaction as an argument everywhere, Zones could work. But is it correct that you want to do something like this?

1. Start transaction.
2. Wait for other events, e.g. user interaction.
3. Commit transaction.

In general I wouldn't recommend that - transactions work better when used for specific short-lived operations. The "scoping" helps to ensure that transactions are always committed or rolled back, and errors are handled properly. I may be able to help better if you explain in more detail what you're trying to do with the transactions.

That said, you can use Completers to achieve something like that. The code below is not tested, and doesn't properly handle errors, but may give a starting point.

```dart
SqliteWriteContext? currentTx;
Completer<void>? _currentTxCompleter;

Future<SqliteWriteContext> startTransaction() async {
  assert(_currentTxCompleter == null);

  _currentTxCompleter = Completer();

  var txResult = Completer<SqliteWriteContext>();

  db.writeTransaction((tx) async {
    currentTx = tx;
    txResult.complete(tx);

    // Wait for endTransaction()
    await _currentTxCompleter!.future;
  });

  // Wait for the transaction to start
  return await txResult.future;
}

void endTransaction() {
  assert(_currentTxCompleter != null);
  _currentTxCompleter!.complete();
  _currentTxCompleter = null;
  currentTx = null;
}
```

**Reply by aleripe** [reply]
Oh no, that's something I would never do! :)
I'm trying to convert my app from Sqflite to your library without having to rewrite everything. That's what I do with the current Sqflite implementation:

```
@override
  Future<T> executeInTransaction<T>(Future<T> Function() action) async {
    return await _lock.synchronized<T>(() async {
      final db = await _database;
      if (_transaction != null) {
        return await action();
      }
      return await db.transaction<T>((transaction) async {
        _transaction = transaction;
        final result = await action();
        _transaction = null;
        return result;
      });
    });
  }
```

An example of `action()` can be an insert/update/delete defined like that:

```
Future<int> insert(String table, Map<String, dynamic> values,
      {ConflictAlgorithm? conflictAlgorithm}) async {
    return await _lock.synchronized(() async {
      final ex = await executor;
      return await ex.insert(table, values,
          conflictAlgorithm: ConflictAlgorithm.replace);
    });
  }
```

Each` action()` uses an executor to execute its command, and executor is retrieved like that:

```
Future<DatabaseExecutor> get executor async {
    if (_transaction != null) {
      return _transaction!;
    } else {
      return await _database;
    }
  }
```

This way, if a transaction has already been opened in the current unit of work it will be used, otherwise a new one will be created.

**Reply by rkistner** [reply]
Ok, I think that makes more sense now.

That same pattern should work with very little modification. Changes:
 * `db.writeTransaction` instead of `db.transaction`
 * `SqliteWriteContext` instead of `DatabaseExecutor` for both `get executor` and `_transaction`.

Are there any further issues you're running into with this?

**Reply by aleripe** [reply]
That's what I did at first but then ran into obscure errors about the database being locked or the transaction closed. I think I have to do some simpler testing to grasp the concepts of your library before going into the harder cases. Thank you so much, you have been very helpful!!

**Reply by aleripe** [reply]
Hi Ralf, I managed to get it working with your suggestions. Now the thing I'm noticing is that the app seems slower than the version using Sqflite, while I supposed it was the opposite as your library uses FFI and async calls. Where should I look in your opinion? There is some area that could be the cause? Thanks!

**Reply by rkistner** [reply, marked as answer]
Performance is tricky, and I can't tell why it would be slower without seeing a specific example.

If you were already using `sqflite_common_ffi`, many operations would have similar performance between the two libraries. I did a blog post a while back comparing performance between different libraries - that may give you a starting point for the general performance you should expect:
https://www.powersync.co/blog/flutter-database-comparison-sqlite-async-sqflite-objectbox-isar

Going off those benchmarks:
1. When using individual operations without transactions, sqlite_async is much faster due to using the WAL mode.
2. Many individual insert/update/delete operations (thousands) in a transaction may be slightly faster with sqlite_async, but you'd get much better performance with both if you batch the operations.
3. Batching large numbers of operations should have better performance with sqlite_async that sqflite.
4. Individual select queries should have similar performance.
5. If it's a small number of slow queries, you'll have no meaningful difference in performance between the two - you'd need to optimize the queries instead.
6. sqlite_async has much better support for concurrency. This is relevant if you need to do e.g. background downloads or uploads while querying data in the app.

**Reply by aleripe** [reply]
Thank you so much Ralf!! Thanks to your suggestions I was able to achieve very good performance. I'm now using a singleton class to manage my database connection, but I have yet to understand if this can be a good approach. In the meantime I'm very satisfied with the results! Thanks again

**Comment by aleripe** [comment]
> Ok, I think that makes more sense now.
> 
> That same pattern should work with very little modification. Changes:
> 
> * `db.writeTransaction` instead of `db.transaction`
> * `SqliteWriteContext` instead of `DatabaseExecutor` for both `get executor` and `_transaction`.
> 
> Are there any further issues you're running into with this?

Hi Ralf, is there a way to check if a transaction has been closed (for example for an automatic rollback)? Thanks!

**Reply by rkistner** [reply]
There isn't currently an API available for that. What would you like to use it for?

**Reply by aleripe** [reply]
I save the current transaction to manage nested transactions, but then I have no way of verifying if the current transaction has been closed due to a rollback and then subsequent calls give me an error "the Transaction has been closed". I'd like to create the current transaction if it is not yet available or if the current has been closed. Does this make sense to you? Thanks!

**Reply by rkistner** [reply]
The best way to do that is to call this is to use finally / whenComplete on the `writeTransaction` / `readTransaction` call, for example:

```dart
await db.writeTransaction((tx) {
  currentTransaction = tx;
  try {
    // ...
  } finally {
    currentTransaction = null;
  }
});
```

**Reply by aleripe** [reply]
Yes, this is what I was thinking about, but was trying to see if there was something more explicit. Thanks again!

**Reply by rkistner** [reply]
Update: In v0.6.0 you can now check whether a transaction is closed by using a combination of:

1. `tx.closed` to check whether the lock is still held (i.e. the callback function in `db.writeTransaction` hasn't returned yet).
2. `await tx.getAutoCommit()` to check whether the transaction is still active or has been rolled back.

**Reply by aleripe** [reply]
Thanks Ralf! I'll have a look at it
